### ISSUE TITLE
Implement Hierarchical Policy for Strategy Selection

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#51]
## Problem

FuturesOneStepEnv agents converge to **single strategies** (Action 6: 87% usage) and cannot adapt to different market regimes.

**Evidence:**
- Single-action dominance across all reward function iterations
- Cannot handle regime shifts (train 2023-2024 vs test 2025)
- No mechanism to switch strategies based on market conditions

---

## Proposed Solution: Hierarchical Policy

**Two-level decision making:**
1. **Meta-policy**: Chooses high-level strategy (aggressive/conservative/balanced)
2. **Sub-policies**: Choose specific actions within that strategy

### Architecture

```
                    ┌─────────────────┐
                    │  Market State   │
                    │  (observation)  │
                    └────────┬────────┘
                             │
                    ┌────────▼────────┐
                    │  META-POLICY    │
                    │  (strategy net) │
                    └────────┬────────┘
                             │
               ┌─────────────┼─────────────┐
               │             │             │
         ┌─────▼─────┐ ┌────▼─────┐ ┌────▼─────┐
         │AGGRESSIVE │ │BALANCED  │ │CONSERV.  │
         │SUB-POLICY │ │SUB-POLICY│ │SUB-POLICY│
         └─────┬─────┘ └────┬─────┘ └────┬─────┘
               │             │             │
               └─────────────┼─────────────┘
                             │
                      ┌──────▼──────┐
                      │   ACTION    │
                      └─────────────┘
```

---

### Strategy Definitions

```python
STRATEGIES = {
    "aggressive": {
        "allowed_actions": [6, 7, 8, 15, 16, 17],  # High TP/SL ratios
        "description": "Wide stops, large profits",
        "risk_tolerance": 0.8,
    },
    "balanced": {
        "allowed_actions": [3, 4, 5, 12, 13, 14],  # Medium TP/SL
        "description": "Balanced risk/reward",
        "risk_tolerance": 0.5,
    },
    "conservative": {
        "allowed_actions": [0, 1, 2, 9, 10, 11],  # Tight stops
        "description": "Small stops, quick exits",
        "risk_tolerance": 0.2,
    },
}
```

### Meta-Policy Network

```python
class MetaPolicy(nn.Module):
    """High-level policy that chooses strategy."""
    
    def __init__(self, obs_dim: int, num_strategies: int = 3):
        super().__init__()
        self.network = nn.Sequential(
            nn.Linear(obs_dim, 128),
            nn.ReLU(),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, num_strategies)
        )
    
    def forward(self, obs: torch.Tensor) -> torch.Tensor:
        """Return strategy logits."""
        return self.network(obs)
```

### Sub-Policy Networks

```python
class SubPolicy(nn.Module):
    """Strategy-specific action selection."""
    
    def __init__(self, obs_dim: int, allowed_actions: List[int]):
        super().__init__()
        self.allowed_actions = allowed_actions
        self.num_actions = len(allowed_actions)
        
        self.network = nn.Sequential(
            nn.Linear(obs_dim, 128),
            nn.ReLU(),
            nn.Linear(128, self.num_actions)
        )
    
    def forward(self, obs: torch.Tensor) -> torch.Tensor:
        """Return action logits for allowed actions only."""
        return self.network(obs)
```

### Hierarchical Actor

```python
class HierarchicalActor(nn.Module):
    """Combines meta-policy and sub-policies."""
    
    def __init__(self, encoder, strategies: Dict):
        super().__init__()
        self.encoder = encoder  # Shared encoder
        self.meta_policy = MetaPolicy(encoder.output_dim, len(strategies))
        
        # Create sub-policy for each strategy
        self.sub_policies = nn.ModuleDict({
            name: SubPolicy(encoder.output_dim, config["allowed_actions"])
            for name, config in strategies.items()
        })
        
        self.strategies = strategies
        self.strategy_names = list(strategies.keys())
    
    def forward(self, obs: TensorDict) -> Tuple[torch.Tensor, torch.Tensor]:
        # Encode observation
        features = self.encoder(obs)
        
        # Meta-policy selects strategy
        strategy_logits = self.meta_policy(features)
        strategy_dist = Categorical(logits=strategy_logits)
        strategy_idx = strategy_dist.sample()
        
        # Sub-policy selects action within strategy
        strategy_name = self.strategy_names[strategy_idx]
        sub_policy = self.sub_policies[strategy_name]
        action_logits = sub_policy(features)
        
        # Map sub-policy action to global action space
        action_dist = Categorical(logits=action_logits)
        local_action = action_dist.sample()
        global_action = self.strategies[strategy_name]["allowed_actions"][local_action]
        
        return global_action, strategy_idx
```

---

### Option A: Joint Training
Train meta-policy and sub-policies simultaneously with PPO.

```python
# Reward shaping to encourage strategy diversity
reward = extrinsic_reward + 0.1 * strategy_entropy_bonus
```

### Option B: Staged Training
1. **Phase 1**: Train sub-policies independently (curriculum per strategy)
2. **Phase 2**: Freeze sub-policies, train meta-policy
3. **Phase 3**: Fine-tune all together

### Option C: Market-Regime Conditioning
Train meta-policy to switch based on market features:

```python
# Add regime features to observation
regime_features = {
    "volatility": compute_volatility(prices),
    "trend": compute_trend(prices),
    "volume_ratio": current_volume / avg_volume,
}

# Meta-policy learns: "In high volatility, use conservative"
```

---

### Success Metrics

| Metric | Baseline | Hierarchical Target |
|--------|----------|---------------------|
| Single action dominance | 87% | < 40% per strategy |
| Strategy switches per episode | 0 | 3-5 |
| Adaptation to regime shifts | None | Automatic |
| Action entropy | 0.48 | > 1.5 |
| Test Sharpe ratio | -5.77 | > 0.5 |

### Behavioral Improvements

1. **Regime adaptation**: Switch from aggressive (trending) to conservative (choppy)
2. **Reduced overfitting**: Different strategies for different market conditions
3. **Interpretability**: Can analyze which strategy used when and why
4. **Robustness**: Less catastrophic failure (multiple strategies = fallbacks)

---

## Advantages

**vs Flat Policy:**
- ✅ Can adapt to market regimes
- ✅ Forces exploration of different strategies
- ✅ More interpretable (strategy-level decisions)
- ✅ Prevents single-action lock-in

**vs Ensemble:**
- ✅ Single model (more efficient)
- ✅ Learns when to use which strategy
- ❌ More complex to train

---

### Week 1: Architecture
- [ ] Implement `MetaPolicy`, `SubPolicy`, `HierarchicalActor`
- [ ] Define strategy configurations
- [ ] Integrate with PPO loss
- [ ] Add strategy-level logging

### Week 2: Training Pipeline
- [ ] Implement joint training approach
- [ ] Add strategy entropy bonus
- [ ] Log meta-policy decisions
- [ ] Visualize strategy switches over episodes

### Week 3: Market Regime Features
- [ ] Add volatility/trend/volume features
- [ ] Train meta-policy with regime conditioning
- [ ] Evaluate regime-specific strategy performance
- [ ] Compare vs baseline and flat exploration approaches

---

## Configuration Example

```yaml
# config.yaml
model:
  type: "hierarchical"
  
  meta_policy:
    hidden_dims: [128, 64]
    activation: "relu"
  
  sub_policies:
    aggressive:
      allowed_actions: [6, 7, 8, 15, 16, 17]
      hidden_dims: [128]
    balanced:
      allowed_actions: [3, 4, 5, 12, 13, 14]
      hidden_dims: [128]
    conservative:
      allowed_actions: [0, 1, 2, 9, 10, 11]
      hidden_dims: [128]
  
  training:
    strategy_entropy_bonus: 0.1
    joint_training: true
```

---

## Monitoring & Debugging

Track these metrics:

```python
metrics = {
    "train/strategy_entropy": strategy_dist.entropy().mean(),
    "train/strategy_0_usage": (strategy_idx == 0).float().mean(),
    "train/strategy_1_usage": (strategy_idx == 1).float().mean(),
    "train/strategy_2_usage": (strategy_idx == 2).float().mean(),
    "train/strategy_switches_per_episode": count_switches(),
    "eval/strategy_vs_market_regime": correlation_analysis(),
}
```

---

### Option D: Continuous Strategy Space
Instead of discrete strategies, use continuous embedding:

```python
class ContinuousMetaPolicy(nn.Module):
    def forward(self, obs):
        # Output 2D strategy embedding: (risk_level, aggressiveness)
        strategy_embedding = self.network(obs)
        return strategy_embedding

# Sub-policy conditioned on embedding
class ConditionedSubPolicy(nn.Module):
    def forward(self, obs, strategy_embedding):
        combined = torch.cat([obs_features, strategy_embedding], dim=-1)
        return self.network(combined)
```

### Option E: Learned Strategy Discovery
Don't predefine strategies - let agent learn them:

```python
# Use VQ-VAE or clustering to discover strategy prototypes
# Agent learns optimal number and type of strategies
```

---

## References

**Hierarchical RL:**
- Sutton et al. (1999) - "Between MDPs and semi-MDPs: A framework for temporal abstraction"
- Bacon et al. (2017) - "The Option-Critic Architecture"
- Nachum et al. (2018) - "Data-Efficient Hierarchical Reinforcement Learning"

**Trading-Specific:**
- Théate & Ernst (2021) - "An application of deep reinforcement learning to algorithmic trading"

---

## Related Issues

- #48 - Curiosity (can combine for sub-policy exploration)
- #49 - Curriculum (can train sub-policies separately)
- #50 - Entropy scheduling (helps meta-policy exploration)
